cylc/flow/task_pool.py

```diff
@@ -1977,34 +1990,87 @@ def set_prereqs_and_outputs(
                     f" {repr_flow_nums(itask.flow_nums, full=True)}"
                 )
                 continue
-            self.merge_flows(itask, flow_nums)
+
             if prereqs:
-                self._set_prereqs_itask(itask, prereqs, flow_nums)
+                set_all, prereqs_valid = (
+                    self._get_prereq_params(prereqs, itask.tdef, itask.point)
+                )
+                if not (set_all or prereqs_valid):
+                    continue
+                self.merge_flows(itask, flow_nums)
+                self._set_prereqs_itask(itask, prereqs_valid, set_all)
+                no_op = False
             else:
+                # Outputs (may be empty list)
                 # Spawn as if seq xtrig of parentless task was satisfied,
                 # with associated task producing these outputs.
+                self.merge_flows(itask, flow_nums)
                 self.check_spawn_psx_task(itask)
                 self._set_outputs_itask(itask, outputs)
+                no_op = False
 
-        # Spawn and set inactive tasks.
         if not flow:
             # default: assign to all active flows
             flow_nums = self._get_active_flow_nums()
+
+        # Spawn and set inactive tasks.
         for tdef, point in inactive_tasks:
             if prereqs:
+                set_all, prereqs_valid = (
+                    self._get_prereq_params(prereqs, tdef, point)
+                )
+                if not (set_all or prereqs_valid):
+                    continue
                 self._set_prereqs_tdef(
-                    point, tdef, prereqs, flow_nums, flow_wait)
+                    point, tdef, prereqs_valid, flow_nums, flow_wait, set_all)
+                no_op = False
             else:
+                # Outputs (may be empty list)
                 trans = self._get_task_proxy_db_outputs(
                     point, tdef, flow_nums,
                     flow_wait=flow_wait, transient=True
                 )
                 if trans is not None:
                     self._set_outputs_itask(trans, outputs)
+                    no_op = False
 
-        if self.compute_runahead():
+        if not no_op and self.compute_runahead():
             self.release_runahead_tasks()
 
+    def _get_valid_prereqs(
+            self, prereqs: List[str], tdef: 'TaskDef', point: 'PointBase'
+    ) -> 'Iterable[Tokens]':
+        """Validate prerequisite triggers and return associated task messages.
+
+        To set prerequisites, the user gives triggers, but we need to use the
+        associated task messages to satisfy the prerequisites of target tasks.
+
+        Args:
+            prereqs:
+                list of string prerequisites of the form "point/task:output"
+        Returns:
+            set of tokens {(cycle, task, task_message),}
+
+        """
+        valid: 'Set[Tuple[str, str, str]]' = set()
+        for pre in tdef.get_prereqs(point):
+            valid.update(list(pre.get_tuples()))
```

Comment:
```suggestion
        valid = {pre.get_tuples() for pre in tdef.get_prereqs(point)}
```

Reply:
Have to unpack the returned lists of tuples:
```python
valid = {key for pre in tdef.get_prereqs(point) for key in pre.keys()}
```